### ISSUE TITLE
Transcribe the octal listing

### ISSUE THREAD

**Issue by thewonderidiot** [virtualagc/virtualagc#884]
This should be nice and easy this time around, since we've got nice dark numbers on a friendly paper!

**Comment by thewonderidiot** [comment]
Prepared octal images produced with octopus: 
[sunburst37.zip](https://github.com/virtualagc/virtualagc/files/1030956/sunburst37.zip)

**Comment by rburkey2005** [comment]
Got 'em!

**Comment by thewonderidiot** [comment]
Woooooow, Sunburst 37 is mad about something. If you run it in yaAGC, it comes up stuck in an infinite reboot loop. If you run with --inhibit-alarms, you can use RSET to turn off the RESET lamp but PINBALL is non-functional.

I popped it in to the hardware sim to see what the heck was going on, and it looks like its infinite loop is happening before a RELINT instruction is ever executed, which explains why PINBALL doesn't work. So we're getting a RUPT LOCK alarm just like we did for Aurora 12, except this one is for no interrupts happening at all instead of getting stuck in an interrupt. However, unlike Aurora 12, it's not self-correcting.

Here's what the waveform trace looks like:
![shepatin_ruptlock](https://cloud.githubusercontent.com/assets/94056/26526013/0c6f6ac0-4328-11e7-83ad-8ea00bdc42ab.png)

The infinite loop is clearly visible, since it regularly sets FB to 01.

I'm guessing it's expecting some sort of erasable load or something... I'll see if I can pinpoint the head of the loop.

**Comment by thewonderidiot** [comment]
Okay, so it turns out it's a massive pain to get Sunburst 37 going with a blank slate. The problem seems to mostly lie in the fact that at the time of Sunburst 37, they weren't using "real" restarts, but rather "fake" restarts through FAKESTRT.

After the computer starts running at address 4000, control is transferred to the GOPROG routine, as per usual. That's the end of the usual, though -- GOPROG begins with a call to this FAKESTRT routine and an interesting notice:
```
GOPROG    TC     BANKCALL   * * * SUBJECT TO A HUGE CHANGE * * *
          CADR   FAKESTRT
          INCR   REDOCTR
```
FAKESTRT is a short little routine that raises a program alarm to show that it occurred, and then jumps to FORGETIT:
```
FAKESTRT  INCR   REDOCTR
          TC     ALARM
          OCT    0316       FAKESTRT ALARM

          TCF    FORGETIT
```

But the call to ALARM has some interesting consequences. It jumps to PROGLARM, which then falls through to MULTEXIT.
```
MULTEXIT  CAF    PRIO37
          TC     NOVAC
          EBANK= FAILREG
          2CADR  LARMLARM
```
We're already in a whole bunch of trouble -- we're calling into the Executive to start a job before any initialization at all has been performed. And the code path to get to this point is completely unavoidable.

So control is transferred to NOVAC in the Executive, which does a bit of setup then jumps to NOVAC2. The first thing it tries to do is find an unused core set, which is defined as having a priority of negative zero. Here's where things start to go really wrong.
```
NOVAC2    CAF    ZERO       NOVAC ENTERS HERE.  FIND A CORE SET.
          TC     LOCCTR                          
          CAF    NO.CORES   SEVEN SETS OF ELEVEN REGISTERS EACH.
NOVAC3    TS     EXECTEM2                        
          INDEX  LOCCTR                          
          CCS    PRIORITY   EACH PRIORITY REGISTER CONTAINS -0 IF
          TCF    NEXTCORE   THE CORrESPONDING CORE SET IS AVAILABLE.
NO.CORES  DEC    6                               
          TCF    NEXTCORE
```
Since the core sets haven't been initialized, the PRIORITY of each is positive zero. Thus, the `CCS PRIORITY` operates on positive zero, and skips one instruction, landing on NO.CORES, which is a constant storing the total number of core sets, and *never* meant to be executed. Since there are 6 core sets, NO.CORES gets executed as an EXTEND. This turns the next `TCF NEXTCORE` instruction into `BZF NEXTCORE`. Miraculously, since the `CCS PRIORITY` left us with +0 in A, this accomplishes the same thing `TCF NEXTCORE` would have.

Since we got lucky, the Executive will successfully check all six core sets to see if any are available. It won't find any, however, since all PRIORITYs are +0. It will thus try to raise the famous 1202 alarm, announcing that no core sets are available.
```
NEXTCORE  CAF    COREINC                         
          ADS    LOCCTR                          
          CCS    EXECTEM2                        
          TCF    NOVAC3                          
          TC     ABORT      NO CORE SETS.
          OCT    1202                            
```

Raising another program alarm through ABORT takes us back into the same alarm code that the earlier ALARM took us to, and it exits the same way -- through MULTEXIT. MULTEXIT will again attempt to start a job, to which the Executive will again raise a 1202, which will then cause MULTEXIT to try to start another job, which will raise another 1202, ad infinitum. This is the loop visible in the waveform I posted above.

So to have any chance of running Sunburst 37, it looks like at least one of the core sets needs to have its PRIORITY set to negative 0. To avoid any other possible weirdness with the `CCS PRIORITY` above, let's use core set 0, the first one checked. Its PRIORITY is located at erasable address 0157, so:
```
0157 = 077777
```
This isn't enough, however. Core set 0 has special handling, and the following code is run almost immediately after core set 0 is found to be available:
```
SPECTEST  CCS    NEWJOB     SEE IF ANY ACTIVE JOBS WAITING (RARE).
          TCF    SETLOC     MUST BE AWAKENED BUT UNCHANGED JOB.
          TC     CCSHOLE
          TC     CCSHOLE
```
With our all-positive-zero initial erasable memory, NEWJOB is +0 and thus `CCS NEWJOB` again takes us to a place we should never hit. This time it's protected by a `TC CCSHOLE`. CCSHOLE is in the alarm logic again, and raises a program alarm, which takes us through MULTEXIT, which tries to start another job. We already took core set 0 and the rest are unitialized, so a 1202 is raised, and we get into the above loop. Therefore, NEWJOB needs to be set to -0 as well. So to even make it past FAKESTRT, we need:
```
0067 = 077777
0157 = 077777
```

This lets Sunburst 37 run for a bit longer, and make it through the rest of standard initialization. But it's still not enough -- it goes for about 70ms before getting itself into a TC trap at address 0.
![tctrap](https://cloud.githubusercontent.com/assets/94056/26564026/0b19da44-4499-11e7-86dc-8112e61c492b.png)

This is also an infinite reboot loop (two iterations of it are visible in the picture above). The specific sequence immediately preceding the TC 0 loop isn't important, because things went awry a bit earlier. At some point during the restart process, Sunburst 37 decides to call DOFSTART -- a totally reasonable thing to do that on most other ropes would fully initialize everything. Not so here though. DOFSTART executes through to ENDRSTRT, which does this as one of its last actions:
```
          CAF    PRIO30     THIS IS USED ONLY FOR SPECIAL ENEMIZE.
          INHINT            $$$ REMOVE IF FULL RESTARTABILITY ***
          TC     FINDVAC
          EBANK= FLUSHREG
          2CADR  FLUSHLOC
```

The FLUSHLOC job starts running shortly thereafter. It's short and sweet:
```
FLUSHLOC  EXTEND            GO TO SPECIAL ENEMA LOC FOR ROM
          DCA    FLUSHREG
          DXCH   Z          USUALLY THIS WILL BE AN ENDOFJOB
```

FLUSHREG is not, however, the 2CADR of ENDOFJOB like the comments suggest. Nothing has initialized it at this point, so the DCA FLUSHREG loads 0 into both A and L. The DXCH Z then does something weird that I hadn't considered before. The 0 in L sets FB to 0 (FLUSHLOC was running out of bank 1). The 0 in A is swapped with Z, putting a 0 into Z and the former value of Z into A. Making Z 0, however, causes the computer to jump to A and execute it as an instruction. Since addresses execute as TCs to themselves, the computer jumps right back to where it was -- except in bank 0 instead of bank 1. Everything goes to hell from there.

To fix this, we can set FLUSHREG to the 2CADR of ENDOFJOB as advertised. We need:
```
0375 = 005605
0376 = 004003
```

With these two erasable settings in place, this is what happens on running Sunburst 37 in yaAGC (sorry for the weird colors):
![sunburst37](https://cloud.githubusercontent.com/assets/94056/26567148/226b37fc-44b6-11e7-8688-8f5f97488cae.gif)

Success! We get notified of a 316 program alarm (the FAKESTRT alarm) and everything seems to be functional. :smile:

In summary, to get Sunburst 37 to run for the first time, the following erasable initializations must be done:
```
0067 = 077777
0157 = 077777
0375 = 005605
0376 = 004003
```
Everything else it is able to set up by itself.

**Comment by thewonderidiot** [comment]
@rburkey2005 -- what's going to be the best way to handle this when we finish Sunburst 37? Provide a ready-made core file?

**Comment by rburkey2005** [comment]
Good job!  I fired off a question about this to Don before reading all the way to the end and seeing that you had solved it.  :smile:  I'm sure he won't recall, but we'll see what he says.

At any rate, I think the best thing would simply be to have a one-off command-line switch in yaAGC (say, --initialize-shepatin-0 or --initialize-sunburst-37) that directly loaded these values in a hard-coded way when yaAGC starts up.  I know that that's not very pleasing aesthetically, but if we didn't hard-code it in yaAGC, and had a special core file for this one program version instead, the use of the extra core file would _still_ have to be hard-coded in VirtualAGC.cpp, and that coding would probably be significantly more complex in VirtualAGC than in yaAGC.  Plus, it would require special manipulations to get the core file into the installer, and so on.

**Comment by rburkey2005** [comment]
@thewonderidiot Actually, I'll go ahead and make the yaAGC CLI-option changed I mentioned.

**Comment by rburkey2005** [comment]
@thewonderidiot I've added it (--initialize-sunburst-37) to yaAGC and pushed it, and Sunburst37 seems to run using it, but I suppose you might want to take a look at it just to be sure.

**Comment by thewonderidiot** [comment]
Looks great! That's exactly what I had for local testing, just wrapped in a command-line option. :smile:
